examples/Example3/particle.h

```diff
@@ -0,0 +1,45 @@
+// SPDX-FileCopyrightText: 2020 CERN
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef PARTICLE_H
+#define PARTICLE_H
+
+#include "point.h"
+#include "vector.h"
+
+struct particle {
+	point position;
+	vector velocity;
+	vector acceleration;
+	float charge;
+	float mass;
+	float time;
+	bool alive;
+
+	__host__ __device__ float energy() const
+	{
+		return 0.5 * mass * norm2(velocity);
+	}
+};
+
+void particle_alloc(int nparticles);
+void particle_resize(int nparticles);
+void particle_free();
+
+int particles_alive();
```

Comment:
The `particle.h` and `particle.cc` files are the module where the system of particles being processed is handled as a whole. Indeed, it may make sense to split it into more parts later, as now several functions are accumulated in this module. I would hold that off a little bit, though, because my intention was to split the functionality when the data layout (AoS, SoA, AoSoA) would be more clear. RIght now, using an array of structs was a compromise to get something working quickly. Ultimately, I'd like to have particle components (position, momentum, etc) stored contiguously in different buffers using AoSoA layout with a fixed block size and managing the number of blocks dynamically with some form of allocator. Then CPU/GPU code only need to know about how to process one such block of particles, and not care where they come from.